src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -382,17 +394,22 @@
    in reverse order (to account for dependency relationships)."
   [app-context :- (schema/atom a/TrapperkeeperAppContext)]
   (log/info "Beginning shutdown sequence")
-  (let [{:keys [ordered-services lifecycle-channel lifecycle-worker]} @app-context
+  (let [{:keys [ordered-services shutdown-channel lifecycle-worker]} @app-context
         shutdown-fn (fn [] (doseq [[service-id s] (reverse ordered-services)]
                              (try
                                (run-lifecycle-fn! app-context s/stop "stop" service-id s)
                                (catch Exception e
                                  (log/error e "Encountered error during shutdown sequence")))))]
-    (async/>!! lifecycle-channel {:type :shutdown
-                                  :task-function shutdown-fn})
+    (log/trace "Putting shutdown message on shutdown channel.")
+    (async/>!! shutdown-channel {:type :shutdown
+                                 :task-function shutdown-fn})
     ;; wait for the channel to send us the return value so we know it's done
-    (async/<!! lifecycle-worker)
-    (log/info "Finished shutdown sequence")))
+    (log/trace "Waiting for response to shutdown message from lifecycle worker.")
+    (if (not (nil? (async/<!! lifecycle-worker)))
+      (log/info "Finished shutdown sequence")
+      ;; else, there was already a shutdown in progress and the second one was ignored
+      (log/trace (str "Response from lifecycle worker indicates shutdown already in progress,"
+                      "ignoring additional shutdown attempt.")))))
```

Comment:
The "else" case here is specifically because the `lifecycle-worker` channel is already closed, right?  If so, would you mind updating the comment and/or log message to clarify that?